### PR TITLE
Cura 7271 slow lines individually minimum layer time

### PR DESCRIPTION
**Pull request by rijkvanmanen**
This definitely needs some code quality checks, but the expected behaviour is realized without generating errors.

### CODE REVIEW

File: src/LayerPlan.cpp [hunk 1]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
+
+    double slowestPathSpeed = DBL_MAX;
+    for (GCodePath& path : paths)
+    {
+        total_extrude_time_at_minimum_speed += path.estimates.extrude_time_at_minimum_speed;
+        total_extrude_time_at_slowest_speed += path.estimates.extrude_time_at_slowest_speed;
+        slowestPathSpeed = std::min(path.config->getSpeed().value * path.speed_factor, slowestPathSpeed);
+    }
+
+    if (totalTime < minTime - 0.01 && extrudeTime > 0.0)
```
Comment (rburema) [inline src/LayerPlan.cpp:1468]:
```suggestion
    constexpr double epsilon = 0.01;
    if (totalTime < minTime - epsilon && extrudeTime > 0.0)
```

File: src/LayerPlan.cpp [hunk 2]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
-            if (path.isTravelPath())
-                continue;
-            double speed = path.config->getSpeed() * path.speed_factor * factor;
-            if (speed < minimalSpeed)
-                factor = minimalSpeed / (path.config->getSpeed() * path.speed_factor);
+            // Even at cool min speed extrusion is not taken enough time. So speed is set to cool min speed.
+            for (GCodePath& path : paths)
+            {
+                if (path.isTravelPath())
+                    continue;
```
Comment (rburema) [inline src/LayerPlan.cpp:1477]:
```suggestion
                if (path.isTravelPath())
                {
                    continue;
                }
```

File: src/LayerPlan.cpp [hunk 3]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
+            if (minTime - total_extrude_time_at_minimum_speed - travelTime > 0.1)
+            {
+                extraTime = minTime - total_extrude_time_at_minimum_speed - travelTime;
+            }
         }
-
-        // Only slow down for the minimal time if that will be slower.
-        assert(getExtrudeSpeedFactor() == 1.0); // The extrude speed factor is assumed not to be changed yet
-        if (factor < 1.0)
+        else if (minExtrudeTime >= total_extrude_time_at_slowest_speed && abs(total_extrude_time_at_minimum_speed - total_extrude_time_at_slowest_speed) >= 0.01)
```
Comment (rburema) [inline src/LayerPlan.cpp:1488]:
```suggestion
        else if (minExtrudeTime >= total_extrude_time_at_slowest_speed && std::abs(total_extrude_time_at_minimum_speed - total_extrude_time_at_slowest_speed) >= epsilon)
```

File: src/LayerPlan.cpp [hunk 4]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
         {
-            setExtrudeSpeedFactor(factor);
+            // Slowing down to the slowest path speed is not sufficient, need to slow down further to the minimum speed.
+            // Linear interpolate between total_extrude_time_at_slowest_speed and total_extrude_time_at_minimum_speed
+            double factor = (total_extrude_time_at_minimum_speed - minExtrudeTime) / (total_extrude_time_at_minimum_speed - total_extrude_time_at_slowest_speed);
+            double target_speed = minimalSpeed * (1-factor) + slowestPathSpeed * factor;
```
Comment (rburema) [inline src/LayerPlan.cpp:1493]:
```suggestion
            double target_speed = minimalSpeed * (1.0 - factor) + slowestPathSpeed * factor;
```

File: src/LayerPlan.cpp [hunk 5]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
+            for (GCodePath& path : paths)
+            {
+                if (path.isTravelPath())
+                    continue;
```
Comment (rburema) [inline src/LayerPlan.cpp:1497]:
```suggestion
                if (path.isTravelPath())
                {
                    continue;
                }
```

File: src/LayerPlan.cpp [hunk 6]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
-        {
-            path.estimates.extrude_time *= inv_factor;
-        }
-
-        if (minTime - (extrudeTime * inv_factor) - travelTime > 0.1)
-        {
-            extraTime = minTime - (extrudeTime * inv_factor) - travelTime;
+            // Slowing down to the slowest_speed is sufficient to respect the minimum layer time.
+            // Linear interpolate between extrudeTime and total_extrude_time_at_slowest_speed
+            double factor = (total_extrude_time_at_slowest_speed - minExtrudeTime) / (total_extrude_time_at_slowest_speed - extrudeTime);
```
Comment (rburema) [inline src/LayerPlan.cpp:1508]:
```suggestion
            const double factor = (total_extrude_time_at_slowest_speed - minExtrudeTime) / (total_extrude_time_at_slowest_speed - extrudeTime);
```

File: src/LayerPlan.cpp [hunk 7]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
+            for (GCodePath& path : paths)
+            {
+                if (path.isTravelPath())
+                    continue;
```
Comment (rburema) [inline src/LayerPlan.cpp:1512]:
```suggestion
                if (path.isTravelPath())
                {
                    continue;
                }
```

File: src/LayerPlan.cpp [hunk 8]
```diff
@@ -1514,12 +1525,28 @@ TimeMaterialEstimates ExtruderPlan::computeNaiveTimeEstimates(Point starting_pos
     TimeMaterialEstimates ret;
     Point p0 = starting_position;
 
+    double min_path_speed = fan_speed_layer_time_settings.cool_min_speed;
+
+    //TODO: This slowest_path_speed is causing warnings for empty layers and looks a bit uckly to me anyway.
+    //      Maybe it can also be returned, because it is now calculated again in the forceMinimalLayerTime function.
+    double slowest_path_speed = DBL_MAX;
```
Comment (rburema) [inline src/LayerPlan.cpp:1532]:
```suggestion
    double slowest_path_speed = std::numeric_limits<double>::max();
```

**Review by rburema — CHANGES_REQUESTED** [review]
Hi!

Didn't get to actually testing this yet, but I do have a ton of remarks/suggestions/observations about code-style.

(Though actually there are only 14 messages ... the rest is because you can't add a code suggestion if you've already made the comment -- so I'd have to cancel the comment to make the code suggestion -- if I could make one I just added it instead.)

Another thing, if you want to use the `numeric_limits` suggestions, you may have to import the `<limits>` header. (In general a good moment to emphasise that I didn't test my own suggestions yet either -- use with care.)

Lastly, the code suggestions are not exhaustive everywhere to prevent (excessive) spam -- like the magic values, 'const'ness, and floating point notation (`1` -> `1.0`).

cheers, Remco.

File: include/pathPlanning/TimeMaterialEstimates.h [hunk 9]
```diff
@@ -17,6 +17,8 @@ class TimeMaterialEstimates
     friend class ExtruderPlan; // cause there the naive estimates are calculated
 private:
     double extrude_time; //!< Time in seconds occupied by extrusion
+    double extrude_time_at_slowest_speed; //!< Time in seconds occupied by extrusion assuming paths are printed at slowest path speed
+    double extrude_time_at_minimum_speed; //!< Time in seconds occupied by extrusion assuming paths are printed at minimum path speed
```
[the remark this answers is not in the document]
Comment (rijkvanmanen) [reply include/pathPlanning/TimeMaterialEstimates.h]: Totally agree this is vague, I tried to improve but I am still not sure.

File: src/LayerPlan.cpp [hunk 10]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
+            for (GCodePath& path : paths)
+            {
+                if (path.isTravelPath())
+                    continue;
+                path.speed_factor = minimalSpeed / (path.config->getSpeed() * path.speed_factor);
+                path.estimates.extrude_time /= path.speed_factor;
+            }
+            // Update stored naive time estimates
+            estimates.extrude_time = total_extrude_time_at_minimum_speed;
+            if (minTime - total_extrude_time_at_minimum_speed - travelTime > 0.1)
```
[the remark this answers is not in the document]
Comment (rijkvanmanen) [reply src/LayerPlan.cpp]: Fixed

File: src/LayerPlan.cpp [hunk 11]
```diff
@@ -1466,46 +1454,69 @@ void LayerPlan::spiralizeWallSlice(const GCodePathConfig& config, ConstPolygonRe
...
+            for (GCodePath& path : paths)
+            {
+                if (path.isTravelPath())
+                    continue;
+                double target_speed = slowestPathSpeed * (1-factor) + (path.config->getSpeed() * path.speed_factor) * factor;
+                path.speed_factor = target_speed / (path.config->getSpeed() * path.speed_factor);
+                path.estimates.extrude_time /= path.speed_factor;
+            }
+            // Update stored naive time estimates
+            estimates.extrude_time = minExtrudeTime;
```
[the remark this answers is not in the document]
Comment (rijkvanmanen) [reply src/LayerPlan.cpp:1525]: I agree this is quite some repeating code. But not sure how to write the proper function for this, can you suggest something?

File: src/LayerPlan.cpp [hunk 12]
```diff
@@ -1514,12 +1525,28 @@ TimeMaterialEstimates ExtruderPlan::computeNaiveTimeEstimates(Point starting_pos
...
+
+    //TODO: This slowest_path_speed is causing warnings for empty layers and looks a bit uckly to me anyway.
+    //      Maybe it can also be returned, because it is now calculated again in the forceMinimalLayerTime function.
+    double slowest_path_speed = DBL_MAX;
+    for (GCodePath& path : paths)
+    {
+        slowest_path_speed = std::min(slowest_path_speed, path.config->getSpeed().value * path.speed_factor);
+    }
+
     bool was_retracted = false; // wrong assumption; won't matter that much. (TODO)
```
[the remark this answers is not in the document]
Comment (rijkvanmanen) [reply src/LayerPlan.cpp:1545]: I also tried to understand, but I decided to leave it

File: src/LayerPlan.cpp [hunk 13]
```diff
@@ -1514,12 +1525,28 @@ TimeMaterialEstimates ExtruderPlan::computeNaiveTimeEstimates(Point starting_pos
...
     for (GCodePath& path : paths)
     {
         bool is_extrusion_path = false;
         double* path_time_estimate;
         double& material_estimate = path.estimates.material;
+
+        //TODO: I don't understand why it is needed to initialize these path estimates at zero. But if I don't do it,
```
[the remark this answers is not in the document]
Comment (rijkvanmanen) [reply src/LayerPlan.cpp]: Thanks for the explanation!

File: src/LayerPlan.cpp [hunk 14]
```diff
@@ -1514,12 +1532,28 @@ TimeMaterialEstimates ExtruderPlan::computeNaiveTimeEstimates(Point starting_pos
     TimeMaterialEstimates ret;
     Point p0 = starting_position;
 
+    double min_path_speed = fan_speed_layer_time_settings.cool_min_speed;
+
+    //TODO: This slowest_path_speed is causing warnings for empty layers and looks a bit uckly to me anyway.
```
Comment (rijkvanmanen) [inline src/LayerPlan.cpp:1537]: Any suggestions for how to fix this todo?

if it was python i would have written: slowest_path_speed = min([path.config->getSpeed().value * path.speed_factor for path in paths if not path.isTravelPath()])

Comment (rburema) [reply src/LayerPlan.cpp]: I think you forgot to check for `isTravelPath()` in the loop.

I've fixed that, as well as the recalculation. The value is now just a class variable.

I've also used the 'accumulate' function to be a bit closer to what python is doing. There is another way to get even closer (using 'ranges'), but I found that in this case it was more hassle than it's worth. It doesn't matter that much in a simple case like this anyway.

**Review by rburema — APPROVED** [review]

**Pull request merged**